A52-xds-custom-lb-policies.md

```diff
@@ -0,0 +1,315 @@
+
+Regardless of the field that load balancing policy configuration is picked up from, internally `XdsClient` will represent the information in a uniform manner. The CDS update struct will be updated to carry LB policy configuration in JSON form. We move from an enum for the policy selection and a set of `oneof` configuration fields to having a [LoadBalancingConfig](https://github.com/grpc/grpc-proto/blob/ab96cf12ec7ce135e03d6ea91d96213fa4cb02af/grpc/service_config/service_config.proto#L560) JSON object.
+
+The load balancer config portion of the CDS update struct will change to take the following form:
+
+```c
+{
+  ...
+
+  "lb_policy_config": <array of LoadBalancingConfig JSON objects>
+  
+  ...
+}
+```
+
+The XdsClient will transfer the responsibility of converting the xDS configuration to the internal JSON format to a new *xDS LB policy registry*. This new registry will return an error if the xDS load balancing configuration cannot be converted, but it will not validate a converted configuration. It will be the responsibility of the XdsClient to validate that the converted configuration. It will do this by having the gRPC LB policy registry parse the configuration.
+
+The CDS resource is rejected and a NACK returned to the control plane in either of the following cases:
+
+- xDS LB policy registry cannot convert the configuration and returns an error 
+- gRPC LB policy registry cannot parse the converted configuration.
+
+### xDS LB Policy Registry
+
+This new registry will maintain a set of converters that are able to map from the xDS LoadBalancingPolicy message to the internal gRPC JSON format. The conversion process will be recursive so that e.g. the conversion of the new xDS WRR Locality policy (see below) will call into the registry to convert the child policies.
+
+The implementation should not allow excessive recursion. Configurations that require more than 16 levels of recursion are considered invalid and should result in a NACK response.
+
+The registry will provide the following function:
+
+```javascript
+// Converts an xDS cluster load balancing policy message to the internal
+// PolicyConfig JSON string.
+//
+// An error/exception will be returned if the message type
+// is not recognized, or if policy configuration is invalid in a way that
+// can't be determined by looking at the PolicyConfig output. 
+string convertToServiceConfig(envoy.config.cluster.v3.LoadBalancingPolicy loadBalancingPolicy)
+```
+
+This function will iterate over the list of `Policy` messages in `LoadBalancingPolicy`, attempting to convert each one to gRPC form, stopping at the first supported policy. The `Policy` message contains a `TypedExtensionConfig` message with the configuration information. `TypedExtensionConfig` in turn uses an Any typed `typed_config` field to store policy configuration of any type. This `typed_config` field is used to determine both the name of a policy and the configuration for it, depending on its type:
```

Comment:
So, for cases where we there are multiple policies in the list -
for example, [policyA, policyB], let's say policyB would run into an error if we tried to convert it to its JSON form, we would still accept this configuration since policyA was understood and we didn't feel the need to verify policyB.

The example can also be modified to say that policyA has troubles with JSON conversion but it is not supported, thereby being accepted by the client. In the future, the client gets updated to support policyA and it now sees a problem with the JSON conversion resulting in the new and updated client NACKing this configuration.

Is my understanding correct?

Reply:
The first scenario you describe is correct. As long as the list of policies has one policy that is both supported by the client and parseable by the xds client, the configuration will be accepted.

In the second one, policyA would not be accepted since it is not supported by the client. The configuration as a whole would only be ACKed if there was another supported  policy in the list with a parseable config. If the client gets updated later to support policyA, it would still not be used if the config was invalid and the client would fall back to any other policy the config might have.